helix-gateway/src/main/java/org/apache/helix/gateway/util/PollChannelUtil.java

```diff
@@ -0,0 +1,131 @@
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.health.v1.HealthCheckRequest;
+import io.grpc.health.v1.HealthCheckResponse;
+import io.grpc.health.v1.HealthGrpc;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PollChannelUtil {
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+  private static final Logger logger = LoggerFactory.getLogger(PollChannelUtil.class);
+
+  // return  <grpc health stub, ManagedChannel> pair
+  // ManagedChannel need to be shutdown when the connection is no longer needed
+  public static Pair<HealthGrpc.HealthBlockingStub, ManagedChannel> createGrpcChannel(String endpointPortString) {
+    String[] endpointPort = endpointPortString.split(":");
+    ManagedChannel channel =
+        ManagedChannelBuilder.forAddress(endpointPort[0], Integer.parseInt(endpointPort[1])).usePlaintext().build();
+
+    return new ImmutablePair<>(HealthGrpc.newBlockingStub(channel), channel);
+  }
+
+  /**
+   * Send Unary RPC to the gRPC service to check the health of the container. Could be liveness or readiness depends on input.
+   * @param service one of "readiness" or "liveness"
+   *                https://github.com/kubernetes/kubernetes/issues/115651
+   * @return
+   */
+  public static boolean fetchLivenessStatusFromGrpcService(String service, HealthGrpc.HealthBlockingStub healthStub) {
+    HealthCheckRequest request = HealthCheckRequest.newBuilder().setService(service).build();
+    HealthCheckResponse response = healthStub.check(request);
+    return response.getStatus() == HealthCheckResponse.ServingStatus.SERVING;
+  }
+
+  /**
+   * Flush the current assignment map to a file. The whole file is re-written every time.
+   */
+  public static void flushAssignmentToFile(String targetAssignment, String filePath) {
+    try (FileWriter fileWriter = new FileWriter(filePath)) {
+      fileWriter.write(targetAssignment);
+    } catch (IOException e) {
+      logger.warn("Failed to write to file: " + filePath, e);
+    }
+  }
+
+  /**
+   * read current state from a file, compare with in memory current state, update the in memory current state and return diff.
+   * Current state file format: {"cluster1" : { "instance_1" : { "resource1" : {"shard1” : “online" }}}}
+   */
+  public static Map<String, Map<String, Map<String, Map<String, String>>>> readCurrentStateFromFile(String filePath) {
+    try {
+      // read from file path
+      File file = new File(filePath);
+      return objectMapper.readValue(file,
+          new TypeReference<Map<String, Map<String, Map<String, Map<String, String>>>>>() {
+          });
+    } catch (IOException e) {
+      logger.warn("Failed to read from file: " + filePath, e);
```

Comment:
Remove the exception. This is recurrent logging. Be safe with the log message length.